### ISSUE TITLE
default discovery workflow for compute nodes is hard coded

### ISSUE THREAD

**Issue by heckj** [RackHD/RackHD#249]
Following up from Slack conversation - 

the default discovery workflow is hard coded and a system configuration can't change what workflow is invoked by default for a new compute node. The relevant code in question is https://github.com/RackHD/on-http/blob/master/lib/services/profiles-api-service.js#L125-L135

The system should allow for a configuration to define an alternate base discovery graph

**Comment by benbp** [comment]
@heckj Would a preferred approach, which is possible today, be to just change the definition for the Graph.Discovery workflow via the API? The downside is it will get replaced on a git update, but the upside, I think, is less complexity because we encourage editing workflows rather than adding additional config variables.

**Comment by heckj** [comment]
Good question - that makes sense to me, not sure what the driver was behind Lakshmi's request in Slack

**Comment by benbp** [comment]
@heckj I think this might tie in nicely as more of a driver to implement the "second" stage cataloging step, where we do further cataloging based on a generated SKU definition. I could also see slotting in some space for "just do whatever extra cataloging the user has specified."

**Comment by lakshminv** [comment]
The current flow that i observed is as follows: 
1. rackHD discovers a compute node
2. Node tries to PXE boot and fails.
3. apply "bootSettings" to discovered node.
4. reboot node.
5. Node now boots with the kernel and initrd specified in the bootSettings.

My requirement is to automate this discovery process eliminating steps 2 and 4 so that any compute node discovered by my rackhd instance auto boots my custom linux image.

**Comment by benbp** [comment]
@lakshminv the way I would recommend doing this is having a SKU driven workflow that runs as part of our extra SKU workflow stage within discovery, where the SKU driven workflow adds the boot settings data. We *are* missing a task that would be able to make updates to a node document, but at the same time it would be trivial to add. Does this sound like something that would work for you?

**Comment by lakshminv** [comment]
@benbp Yes, I've experimented with the SKU driven workflow and i think it will be good for me now. but the problem with SKUs is to create the matching rules. for now i can create rules to match the nodes that I already know. But with new nodes i'm still not confident. Is there a way to specify a SKU rule to apply for all compute nodes?

As you mentioned earlier, could you point me to the API used to change the default Graph.Discovery workflow?
But in the long run, I feel having the default discovery workflow as a configurable parameter would help in many ways.

**Comment by benbp** [comment]
@lakshminv You could just use a catch-all regex to match all nodes, something like

```
rules: [
    "path": "dmi.System Information.Manufacturer",
    "regex": "*"
]
```

The API you can use to change the default Graph.Discovery is just the graph definitions API, so you would be modifying the existing discovery workflow.

```
PUT /api/2.0/workflows/graphs
Content-type: application/json
Body: <workflow definition>

# You can get the existing workflow definition via
GET /api/2.0/workflows/graphs/Graph.SKU.Discovery
GET /api/2.0/workflows/graphs/Graph.Discovery
# The Graph.SKU.Discovery workflow runs the Graph.Discovery workflow as a sub-workflow
```

It seems from your listed requirements that you want to just create bootSettings automatically. I think the simplest way is for us to create a task allowing for node document changes, and then you can append this task to the discovery workflow, or just add it to a SKU workflow.

**Comment by lakshminv** [comment]
thanks @benbp. can you please reuse this ticket to create a task allowing for node document changes.

**Comment by benbp** [comment]
@lakshminv sounds good. We've got some outstanding work coming soon for that. I'll add some links in here.

**Comment by pscharla** [comment]
@lakshminv : Found this old issue that was still open.  So a couple things here if you are still in this space:
1) We have work currently going up (PRs up from @VulpesArtificem ) to enable a configurable discovery workflow. PRs here:
https://github.com/RackHD/on-http/pull/543
https://github.com/RackHD/docs/pull/365

2) We have a ticket in our backlog to be able to modify a node document. Was originally for another use case besides "bootSettings" but we had been getting push back there because of having it too wide open. Anyway, now that I have the specific use case I'll try to push it forward. benbp (who is no longer around) had mentioned this use case, but I didn't know the background/details until I found this ticket.
